numpy/_core/src/umath/dispatching.cpp

```diff
@@ -882,13 +889,49 @@ promote_and_get_info_and_ufuncimpl(PyUFuncObject *ufunc,
     }
 
     /* Add this to the cache using the original types: */
-    if (cacheable && PyArrayIdentityHash_SetItem(ufunc->_dispatch_cache,
-            (PyObject **)op_dtypes, info, 0) < 0) {
+    if (cacheable && PyArrayIdentityHash_SetItem(
+                (PyArrayIdentityHash *)ufunc->_dispatch_cache,
+                (PyObject **)op_dtypes, info, 0) < 0) {
         return NULL;
     }
     return info;
 }
 
+#ifdef Py_GIL_DISABLED
+/*
+ * Fast path for promote_and_get_info_and_ufuncimpl.
+ * Acquires a read lock to check for a cache hit and then
+ * only acquires a write lock on a cache miss to fill the cache
+ */
+static inline PyObject *
+try_promote_and_get_info_and_ufuncimpl(PyUFuncObject *ufunc,
+        PyArrayObject *const ops[],
+        PyArray_DTypeMeta *signature[],
+        PyArray_DTypeMeta *op_dtypes[],
+        npy_bool legacy_promotion_is_possible)
+{
+    ((std::shared_mutex *)((PyArrayIdentityHash *)ufunc->_dispatch_cache)->mutex)->lock_shared();
```

Comment:
it would be a little clearer to define a `std::shared_mutex *mutex` local variable and use that here and below so this complicated pointer casting only happens once